Ensure non-intersecting sf-geometries in get_eurostat_geospatial

retostauffer opened the pull request:
Hy there.

Referring to https://github.com/rOpenGov/eurostat/issues/197 where we noticed that `get_eurostat_geospatial()` in some cases returns invalid geometries (Self-intersecting POLYGONS or MULTIPOLYGONS). With respect to https://github.com/rOpenGov/eurostat/issues/197 I have tried to implement a fix.

Unfortunately a pull request against the master branch, if it is preferred to have it against e.g., the development branch let me know, only few lines of changes to make.

**Changes**

* `get_eurostat_geospatial()` has an additional argument `make_valid = FALSE` (current default).
    * If `output_class == "sf" & !make_valid`: A warning will be thrown; Causes warnings in several places (e.g., the examples in the function manual as, by default, `make_valid = FALSE`).
    * If `output_class == "sf" & make_valid`: Once applying `st_buffer(shp, 0)`; removes the self-intersection problem and ensures that the resulting geometries are valid.
* Further details given in the `@details` section; Small example added in the `@example` section.

**Notes/remarks:**

* I originally suggested to (i) check if there are invalid (self-intersecting) geometries, if so apply `st_buffer()`. Creates unnecessary overhead and increases maintenance, thus suggesting to always apply `st_buffer()`. Microbenchmark: average time for the example used on my laptop ca. 20 milliseconds.
* Self-intersecting geometries do not occur in all data sets! Using `1:60`, nuts level `2` for year `2013` in the example (`?get_eurostat_geospatial`) as it is one of the situations where the problem occurs. 

Feel free to come back to me if I can improve the merge request or if something is unclear or should be changed.

All best,
R

antagomir commented on R/get_eurostat_geospatial.R:
```diff
@@ -25,13 +25,23 @@
+#' This can cause problems, e.g., when using these geometries as input argument
+#' to \code{sf::st_interpolate_aw()}). \code{make_valid = TRUE} makes sure that
+#' only valid (multi-)polygons are returned, example included below.
 #' @author Markus Kainu <markuskainu@gmail.com>
```
Feel free to list yourself as the second author of this function if you prefer.

retostauffer replied:
Thanks for the quick reply @antagomir ; I took a short break and had an idea to improve my approach to avoid unnecessary warnings, working on it. Reason: not all combinations of resolution, nuts_area, and year have this issue. Currently running a test on all combinations to figure out how often it happens and if it makes sense to include my new idea.

I was thinking of adding tests; `get_eurostat_geospatial()`, so far, has no testthat tests wherefore I decided not to add any, happy to add few, trying to update my repos (should update the pull request, right?).

Thx in advance,
R

antagomir replied:
Yes - if you could update the PR with new unit tests that would be awesome!

antagomir reviewed (commented):
Seems good overall, I requested another review from @muuankarski 

Two comments: 
- Would it be feasible to add unit tests in order to guarantee correct function?
- You could add yourself in contributor role (ctb) in DESCRIPTION file?

antagomir commented:
Btw can you confirm when the changes can be considered "ready", we will then have another look.

retostauffer commented:
Work in progress! Worked on a series of tests for for `get_eurospatial_geospatial(output_format = "sf", ...)` (`"sf"` only) and updated the sanity checks for this specific function.

Additional check if `st_buffer()` is required seems to be obsolete as it happens in most of the cases. Hope that I can extend the tests this week to ensure covering most of the function.

retostauffer commented:
Added a series of tests in [tests/test_get_eurostat_geospatial.R](https://github.com/retostauffer/eurostat/blob/master/tests/testthat/test_get_eurostat_geospatial.R) covering 92 percent of the code. 

* Slightly changed the sanity checks of input arguments and added missing one (`nuts_levels`).
* Adjusted conditional execution in few places.
* Modified few commands to require less detailed tests (e.g., `shp <- eval(...)` for `nuts_level != "all"`).
* CI now succeeds for macOS-latest (not sure why; not changed anything in particular).

At first did not wanted to change the content of the function, but makes some things a bit easier/more streamlined and requires less tests. I would consider this as 'done', feel free to implement (or stash) the changes.

antagomir commented:
All checks pass and special mention on the nice coding style!

I would just like to see a confirmation from @muuankarski before we merge.

retostauffer commented:
Just noticed that this pull request is still open; anything I can help with or clarify?

muuankarski reviewed (approved):
My apologies for not reacting on this on time. Very elegant solution! The only change I would make is to add row breaks in example in `get_eurostat_geospatial()`-function on line 60.

antagomir commented:
Yes, sorry that this took time. I think this is ready for merging, can you add the small row breaks to PR before we merge?

Then I think we could at the same go make a new CRAN release. I can handle that.

antagomir commented:
I can add the row breaks, so we get this completed.

antagomir commented:
I merged, did some other minor updates. Now testing with devtools/win_devel. 
Also added @retostauffer to DESCRIPTION file with "ctb" role.
Submitting updated version to CRAN soon.
Thanks!

retostauffer commented:
My apologies not to react earlier, was busy coding other stuff over the weekend. Thanks a lot your fix and nice to hear it made it - my Master student will be happy :), very much appreciated.

antagomir commented:
Noprob, we were busy too but happy we did it. Now we still need to overcome the CRAN battle - if there are others who should be acknowledged just let us know.
